### PR TITLE
switch from git2r to gert

### PR DESCRIPTION
**Pull request by hansvancalster**
Most functions and corresponding tests now use `gert` instead of `git2r` functions. 
This should allow both https- and ssh-protocol for communication with Git and GitHub.

In `check_description.R` on lines 88-91 I have left some lines of code which still use `git2r` functions because I could not find a `gert` alternative (yet).
Also in `check_filename.R` on line 53, I haven't found a good alternative for `git2r::in_repository()`.

### CODE REVIEW

**Comment by hansvancalster** [comment]
When I run `checklist::check_package()` locally, all is OK and all tests pass. They fail here and I suspect this has to do with GHA workflows specifying `master` instead of `main` as main branch (? - just a feeling)

**Comment by hansvancalster** [comment]
> When I run `checklist::check_package()` locally, all is OK and all tests pass. They fail here and I suspect this has to do with GHA workflows specifying `master` instead of `main` as main branch (? - just a feeling)

On second thought, it's probably because these are breaking changes. Checking the checklist package in this branch with current checklist version on main branch will fail. Locally I check the package with a fresh built of the package that's being checked 🤯

**Comment by hansvancalster** [comment]
> On second thought, it's probably because these are breaking changes. Checking the checklist package in this branch with current checklist version on main branch will fail. Locally I check the package with a fresh built of the package that's being checked 🤯

I just did `remotes::install_github("inbo/checklist") and then ran `checklist::check_package()` and this returned only a couple of lintr issues (which all look like false positives to me; I need to stop can't think straight...):

```
############################################################################################################################
14 linters found.
`styler::style_file()` can fix some problems automatically. 

----------------------------------------------------------------------------------------------------------------------------
6 times "no visible global function definition for ‘git_status’"
2 times "no visible global function definition for ‘git_branch_checkout’"
2 times "no visible global function definition for ‘git_branch_list’"
1 times "no visible global function definition for ‘git_branch_delete’"
1 times "no visible global function definition for ‘git_fetch’"
1 times "no visible global function definition for ‘git_ls’"
1 times "possible error in clean_git(repo = repo, verbose = verbose): unused argument (repo = repo)"

############################################################################################################################

Checking the package revealed some problems.
```

**Comment by ElsLommelen** [comment]
Just from the errors itself, it seems you might have forgotten to add these functions as dependencies.
And last error: probably argument repo doesn't exist in `clean_git()`

**Comment by ElsLommelen** [comment]
Just from the errors itself, it seems you might have forgotten to add these functions as dependencies.
And last error: probably argument repo doesn't exist in `clean_git()`

**Comment by hansvancalster** [comment]
> Just from the errors itself, it seems you might have forgotten to add these functions as dependencies.
> And last error: probably argument repo doesn't exist in `clean_git()`

No that's not the problem. They are all loaded via importFrom and are present in the namespace and package `gert` is in description imports section. The last error is also false: there is an argument repo.

**Comment by florisvdh** [comment]
The first error on GHA, indicated at https://github.com/inbo/checklist/runs/3633245557#step:10:374, seems the cause of all subsequent errors:

```
 ── Error (test_b_clean_git.R:27:3): clean_git with `main` as main branch ───────
Error: Error: src refspec 'refs/heads/main' does not match any existing object
```

I didn't look into the test file (or setup for it), but this looks like branch misnaming indeed, or perhaps worse (not running in a git repo, or forgot to fetch?).

File: NAMESPACE [hunk 1]
```diff
@@ -39,37 +39,33 @@ importFrom(desc,desc)
...
+importFrom(gert,git_info)
+importFrom(gert,git_init)
+importFrom(gert,git_log)
+importFrom(gert,git_ls)
+importFrom(gert,git_push)
+importFrom(gert,git_stat_files)
+importFrom(gert,git_status)
+importFrom(gert,git_tag_create)
+importFrom(gert,git_tag_list)
 importFrom(git2r,branches)
```
Comment (ThierryO) [inline NAMESPACE:63]: There are still `git2r` functions in the `NAMESPACE`. We should either fully switch to `gert` or keep using `git2r`.

File: R/check_description.R [hunk 2]
```diff
@@ -49,39 +50,40 @@ check_description <- function(x = ".") {
     !grepl("^[0-9]+\\.[0-9]+(\\.[0-9]+)?$", version)
   ] -> desc_error
   notes <- character(0)
-  if (length(commits(repo)) > 1) {
-    branch_sha <- vapply(branches(repo, "all"), branch_target, character(1))
-    head_sha <- sha(repository_head(repo))
-    current_branch <- head(names(which(branch_sha == head_sha)), 1)
+  if (length(gert::git_log(ref = gert::git_branch(repo = repo),
```
Comment (ThierryO) [inline R/check_description.R:53]: Don't use the `package::function` notation. Unless really need functions from multiple package with the same name.

File: tests/testthat/test_a_is_workdir_clean.R [hunk 3]
```diff
@@ -2,26 +2,28 @@ test_that("is_workdir_clean", {
   path <- tempfile("clean")
   dir.create(path)
   on.exit(unlink(path, recursive = TRUE), add = TRUE)
-  repo <- init(path)
-  git2r::config(repo = repo, user.name = "junk", user.email = "junk@inbo.be")
+
+  gert::git_init(path = path)
```
Comment (ThierryO) [inline tests/testthat/test_a_is_workdir_clean.R:6]: Here you only need the `package::function` syntax when you need a function that `checklist` does not import. Hence `git2r::config()` but not `git2r::add()`.

File: tests/testthat/test_c_create_package.R [hunk 4]
```diff
@@ -23,6 +23,13 @@ test_that("create_package() works", {
     regexp = sprintf("package created at `.*%s`", package)
   )
 
+  repo <- file.path(path, package)
```
Comment (ThierryO) [inline tests/testthat/test_c_create_package.R:26]: `create_package()` should initialise a git repo. So no need to do that manually.

**Review by ThierryO — CHANGES_REQUESTED** [review]
Please fully switch from `git2r` to `gert`.

**Comment by ThierryO** [comment]
> In `check_description.R` on lines 88-91 I have left some lines of code which still use `git2r` functions because I could not find a `gert` alternative (yet).
> Also in `check_filename.R` on line 53, I haven't found a good alternative for `git2r::in_repository()`.

I don't like to import two packages with very similar functions into the same package.
That  makes things confusing.

**Comment by ThierryO** [comment]
The `gert:::raise_libgit2_error` seems to suggest that the libgit2 linux library is not available.

You need to add `gert` to the [`Dockerfile`](https://github.com/inbo/checklist/blob/eec09a92640f46f7d5a861dff0685bd794a7ce70/Dockerfile#L40). Untill the release with `gert` you'll need to add the required [`aptget`](https://github.com/inbo/actions/tree/master/check_pkg) to the GHA.

File: NAMESPACE (hunk 1 above)
Comment (hansvancalster) [reply NAMESPACE:63]: I know, but I haven't found a solution yet for some of the functions.

**Comment by hansvancalster** [comment]
> The `gert:::raise_libgit2_error` seems to suggest that the libgit2 linux library is not available.
> 
> You need to add `gert` to the [`Dockerfile`](https://github.com/inbo/checklist/blob/eec09a92640f46f7d5a861dff0685bd794a7ce70/Dockerfile#L40). Untill the release with `gert` you'll need to add the required [`aptget`](https://github.com/inbo/actions/tree/master/check_pkg) to the GHA.

From https://github.com/r-lib/gert#installation it seems I need to do the following for Ubuntu Trusty and Xenial:

```
sudo add-apt-repository ppa:cran/libgit2
sudo apt-get update
sudo apt-get install libgit2-dev
```

The [`aptget`](https://github.com/inbo/actions/tree/master/check_pkg) already has `libgit2-dev` and runs on ubuntu-latest. Do you think I need to change something? Something like:

```
on: [push]

name: check-pkg

jobs:
  check_package:
    runs-on: ubuntu-latest
    name: Check package
    steps:
      - uses: inbo/actions/check_pkg@master
        with:
          - path: my_path
          - CODECOV_TOKEN: "some codecov token"
          - ORCID_TOKEN: "my ORCID token"
          - add-apt-repository: ppa:cran/libgit2 #syntax correct? colons in the correct place?
          - apt-get: update #syntax correct? colons in the correct place?
          - apt-get: install libgit2-dev #syntax correct? colons in the correct place?
```

**Comment by ThierryO** [comment]
You need to provide a comma separated list of apt-get package to install. Custom ppa are currently not supported.

```
on: [push]

name: check-pkg

jobs:
  check_package:
    runs-on: ubuntu-latest
    name: Check package
    steps:
      - uses: inbo/actions/check_pkg@master
        with:
          - path: my_path
          - CODECOV_TOKEN: "some codecov token"
          - ORCID_TOKEN: "my ORCID token"
          - apt-get: libgit2-dev
```

File: R/check_description.R (hunk 2 above)
Comment (hansvancalster) [reply R/check_description.R]: Done in 1563e1e

File: tests/testthat/test_a_is_workdir_clean.R (hunk 3 above)
Comment (hansvancalster) [reply tests/testthat/test_a_is_workdir_clean.R]: Done in 1563e1e

**Comment by hansvancalster** [comment]
> ```
> ############################################################################################################################
> 14 linters found.
> `styler::style_file()` can fix some problems automatically. 
> 
> ----------------------------------------------------------------------------------------------------------------------------
> 6 times "no visible global function definition for ‘git_status’"
> 2 times "no visible global function definition for ‘git_branch_checkout’"
> 2 times "no visible global function definition for ‘git_branch_list’"
> 1 times "no visible global function definition for ‘git_branch_delete’"
> 1 times "no visible global function definition for ‘git_fetch’"
> 1 times "no visible global function definition for ‘git_ls’"
> 1 times "possible error in clean_git(repo = repo, verbose = verbose): unused argument (repo = repo)"
> 
> ############################################################################################################################
> 
> Checking the package revealed some problems.
> ```

The issue with the false positive lintrs is gone after 1563e1e (removing redundant gert:: ; which is really strange: apparently lintr throws a bunch of unrelated issues. It probably silently assumes that one does not use function:: in a redundant way?)

File: tests/testthat/test_c_create_package.R (hunk 4 above)
Comment (hansvancalster) [reply tests/testthat/test_c_create_package.R:26]: Done in 5c98cd1

**Comment by hansvancalster** [comment]
> The `gert:::raise_libgit2_error` seems to suggest that the libgit2 linux library is not available.
> 
> You need to add `gert` to the [`Dockerfile`](https://github.com/inbo/checklist/blob/eec09a92640f46f7d5a861dff0685bd794a7ce70/Dockerfile#L40). Untill the release with `gert` you'll need to add the required [`aptget`](https://github.com/inbo/actions/tree/master/check_pkg) to the GHA.

This was not necessary. The problem was that tests failed in GHA because in that environment the main branch was called `master` whereas when I ran the tests locally, the main branch was called `main`. This is now fixed by setting the main branch name programmatically in the tests.

**Comment by hansvancalster** [comment]
One of the tests in `test_b_clean_git.R` still fails in GHA. When I run it locally everything is fine. Can you have a look at this one failing test, because I have no clue what's going on.

File: NAMESPACE (hunk 1 above)
Comment (hansvancalster) [reply NAMESPACE]: Done in d91a2f5

**Comment by hansvancalster** [comment]
Commit a1ac0f2 needs careful review, because I was not sure from the documentation of `git2r::diff.git_tree` if my git shell commands are entirely equivalent. In the code I am not diffing trees but diffing the commits that go with the trees (which should be OK I think?). I've used the `git diff <commit>..<commit> -- [<path>]` syntax explained [here](https://git-scm.com/docs/git-diff#Documentation/git-diff.txt-emgitdiffemltoptionsgtltcommitgtltcommitgt--ltpathgt82308203).

**Review by hansvancalster — COMMENTED** [review]
Good catch 🚀

**Comment by ThierryO** [comment]
@hansvancalster Feel free to add your name as contributor to the `DESCRIPTION`.

**Review by ThierryO — APPROVED** [review]

**Pull request merged**